otherlibs/unix/channels_win32.c

```diff
@@ -60,15 +60,36 @@ static DWORD check_stream_semantics(value handle)
   }
 }
 
+int caml_win32_get_CRT_fd(value handle)
+{
+  int fd;
+  SPIN_WAIT {
+    fd = atomic_load(&CRT_fd_val(handle));
+    if (fd != GETTING_CRT_FD) return fd;
+  }
+}
+
 int caml_win32_CRT_fd_of_filedescr(value handle)
 {
-  if (CRT_fd_val(handle) != NO_CRT_FD) {
-    return CRT_fd_val(handle);
-  } else {
-    int fd = _open_osfhandle((intptr_t) Handle_val(handle), O_BINARY);
-    if (fd == -1) caml_uerror("channel_of_descr", Nothing);
-    CRT_fd_val(handle) = fd;
-    return fd;
+  SPIN_WAIT {
+    int fd = atomic_load(&CRT_fd_val(handle));
+    switch (fd) {
+    case NO_CRT_FD:
+      if (! atomic_compare_exchange_strong(&CRT_fd_val(handle),
+                                           &fd, GETTING_CRT_FD))
+        break; /* try again */
+      fd = _open_osfhandle((intptr_t) Handle_val(handle), O_BINARY);
+      if (fd == -1) {
+        atomic_store(&CRT_fd_val(handle), NO_CRT_FD);
+        caml_uerror("channel_of_descr", Nothing);
+      }
+      atomic_store(&CRT_fd_val(handle), fd);
+      return fd;
+    case GETTING_CRT_FD:
+      break; /* try again */
+    default:
+      return fd;
+  }
   }
```

Comment:
The indention looks wrong here, but I think that may be the GitHub UI doing something weird, as it's suggesting these two braces are unchanged??